doc/design.rst

```diff
@@ -209,3 +209,237 @@ _________________
+
+
+Poetry for Project Management
++++++++++++++++++++++++++++++
+
+While poetry was and is a good choice for exasol project, dependency, build tool etc. "most recently"
+`uv <https://docs.astral.sh/uv/>`_ has surffeced and made big advanced. Looking at uv it addresses additional itches with
+our projects and therefore in the long run it may be a good idea to migrate our project setups to it.
+Use poetry for project, build and depenency management.
+
+
+Code Formatting
++++++++++++++++
+
+**Description:**
+
+Currently we use Black and Isort for code formatting, though running them on a larger code base as pre-commit hooks or such can take quite a bit of time.
+
+**Downsides:**
+
+- Two tools and an aligned configuration of them are required to cleanly and correctly format the codebase.
+- Code needs to be processed at least twice as we apply two individual tools.
+- The performance of Black and Isort is okay but not great compared to other tools.
+
+**Rationale/History:**
+
+- Black and Isort have been used because they are battle-tested and widely used
+- When we opted for Black and Isort, ``ruff`` wasn't "a thing" yet and at best in its early stages.
+- Black and Isort already have been known by most python devs when we where selecting the tools
+
+**Ideas/Solutions:**
+
+As `Ruff <https://docs.astral.sh/ruff/>`_ is fairly stable and also tested and used by many Python projects
+we should consider transitioning to it.
+
+Advantages:
+
+- Well-tested
+- Widely used
+- Excellent performance
+- Single tool for imports and formatting the codebase
+- Simplifies adopting ruff for linting
+
+
+Pylint
+++++++
+
+**Description:**
+We are currently using Pylint instead of Ruff.
+
+**Downsides:**
+
+- Pylint is slower and less usable in pre-commit hooks
+- It is an additional tool, therefore at leas one more processing run of the code is required
+- No LSP (e.g. compare to `ruff lsp`)
+
+**Rationale/History:**
+
+- Well known
+- Pylint provides built-in project score/rating
+- Project score is good for improving legacy code bases which haven't been linted previously
+- Plugin support
+
+**Ideas/Possible Solutions:**
+
+Replacing Pylint with Ruff for linting would provide significant performance improvement. Additionally, Ruff offers an LSP and IDE integrations and is widely used these days. Additionaly there would be an additional synergy if we adopt ruff for formatting the code base.
+
+Transitioning to Ruff requires us to adjust the migration and improvement strategies for our projects:
+
+- Currently, our codebase improvements are guided by scores. However, with Ruff, a new approach is necessary. For example, we could incrementally introduce specific linting rules, fix the related issues, and then enforce these rules.
+
+- The project rating and scoring system will also need modification. One possiblity would be to run Ruff and Pylint in parallel, utilizing Pylint solely for rating and issue resolution while Ruff is incorporated for linting tasks.
```

Comment:
Could Sonar be an alternative for open-source projects? Sonar supports incremental linting, i.e. findings in unmodified existing code are not considered as build breaker.
Sonar also provides code coverage and PR merge gates via branch protection rules.

Reply:
I think Sonar could be used or integrated, but solely depending on a web service for this is not a good choice, in my opinion. That said, Sonar supports input from third-party tools like `ruff` and `pylint`. See [here](https://docs.sonarsource.com/sonarqube-cloud/enriching/external-analyzer-reports/#python).